[RFC] Split pty_process into OS specific files

equalsraf opened the pull request:
~~Added new guard FEAT_PTY_PROCESS to enable/disable the use of pty specific
code - enabled by default for UNIX.~~

**EDIT**: A pty implementation is not available in windows yet. Use the regular libuv_process functions for now, and split pty_process.c into platform specific files.

Ping #810.

justinmk commented on config/config.h.in:
```diff
@@ -62,4 +62,8 @@
 
+#ifdef UNIX
+# define FEAT_PTY_PROCESS
+#endif
```
Can this be driven by CMakeLists.txt instead of checking `#ifdef UNIX` here?

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -65,6 +65,10 @@ foreach(sfile ${NEOVIM_SOURCES})
   if(${f} MATCHES "^(regexp_nfa.c)$")
```
@tarruda What is the reason for the special handling of `regexp_nfa.c` in this cmake file?

equalsraf replied:
regexp_nfa.c is included in regexp.c so that is why we don't put it in the sources var, it still needs to generate headers though.

ZyX-I replied:
I would name such files “foo.c.h”. And @equalsraf is right, this is a “part” of regexp.c and cannot be compiled on its own.

equalsraf commented:
Need to pull in other commits from #810, pty is required for :terminal.

tarruda commented:
:+1: 

FWIW I'd like to eventually support `:terminal` in windows through [winpty](https://github.com/rprichard/winpty)

equalsraf commented:
@tarruda sure, but this is one of the biggest bits blocking a working build in master.

tarruda commented:
@equalsraf wouldn't it be easier to just treat pty processes as libuv processes instead of disabling `:terminal`? For example, in the windows build you could:

- disable compilation of pty_process.c
- add an #ifdef to pty_process.h that
 - `#define PtyProcess LibuvProcess`
 - alias each pty_process function as libuv_process with a #define
- make any necessary adjustments to compile

This is possible because the interface to the terminal module is just a byte stream, so we'd just need to ignore the pty-specific functions such as resize. It may even be possible to spawn programs such as powershell.exe or cmd.exe outside a `:terminal` window(though console-specific features such as line editing won't work properly).

equalsraf commented:
> @equalsraf wouldn't it be easier to just treat pty processes as libuv processes instead of disabling 

Possible, but wouldn't amount to much - I doubt anyone would want to build with terminal and without pty, it would just be a pointless build setting (tests would not work, etc). Also for Windows :terminal has other dependencies (libvterm).

tarruda commented:
> Possible, but wouldn't amount to much - I doubt anyone would want to build with terminal and without pty, it would just be a pointless build setting (tests would not work, etc)

Programs can still send esc codes to stdout/stderr even when these are normal pipes. Some programs like bash will test if stdout is a terminal in order to decide if a prompt should be displayed or if it should send colored output, but other than that a pty behaves a lot like simple pipes(Most distinction lies in the line discipline layer and in the API for querying/setting window sizes, but this is mostly noticeable by the master side of the pty).

So if a program were to send esc codes to Neovim(regardless if it was spawned over a pty connection), it would still see colored output and `:terminal` would work a lot like on unix. This is the concept behind [winpty](https://github.com/rprichard/winpty): It will intercept changes to a hidden console window and generate a stream of esc codes to the terminal emulator.

[mintty](https://code.google.com/p/mintty/) already states in its website that while it works for msys/cygwin bash and other unix shells(compiled to the respective msys/cygwin runtimes), only simple text-only programs that don't make a distinction between terminals and pipes will work unmodified:

> Mintty works on all Windows versions from Windows 2000 onwards. Similarly to other Cygwin/MSYS terminals based on pseudo terminal ("pty") devices, however, mintty is not a full replacement for the Windows Command Prompt. While native console programs with simple text output usually work fine, interactive programs often have problems, although sometimes there are workarounds.

The workaround mentioned in the quote refers to the -i flag for win32 python. That means python.exe would probably behave a lot like its unix counterpart if spawned as `:terminal python.exe -i`.

In fact, I'm starting to think that fully porting `:terminal` to windows is a lot simpler that I initially considered, it might not even require linking against winpty which already provides a wrapper program that can be used directly by mintty(and possibly by `:terminal` too). Just aliasing PtyProcess to LibuvProcess(as I suggested in the previous comment) and prefixing programs with winpty's `console.exe` might be enough.

tarruda commented:
> Also for Windows :terminal has other dependencies (libvterm).

libvterm is also a dependency for unix, and AFAIK it is a pure C99 library without any dependencies on posix API.

equalsraf commented:
@tarruda, ok changed commit to fallback to libuvprocess and reenabled terminal:

- Ended up redefining PtyProcess minus the fd
- Whats the correct defines for pty_process_teardown() and pty_process_close_master()?

> libvterm is also a dependency for unix, and AFAIK it is a pure C99 library without any dependencies on posix API.

MSVC does not support VLAs - I have a working version libvterm for windows building with CMake [here](https://github.com/equalsraf/libvterm-win).

The windows console is visible, but input doesn't seem to reach the console (or at least no updates are visible). And resizes loose the content.

![capture](https://cloud.githubusercontent.com/assets/715238/12469095/dc989fcc-bfe0-11e5-9259-f4c6fc3d76c3.PNG)

tarruda commented:
@equalsraf thanks for trying it out

> Whats the correct defines for pty_process_teardown() and pty_process_close_master()?

Both these functions are irrelevant for libuv processes, so they can be defined as empty macros. Note that `pty_process_close_master()` is only called from process.c when `proc->type == kProcessTypePty`, which won't happen since all process should be spawned from libuv.

> The windows console is visible, but input doesn't seem to reach the console (or at least no updates are visible). And resizes loose the content.

I believe cmd.exe only works correctly when attached to a console window. What happens if you spawn it using winpty `console.exe`? The [documentation](https://github.com/rprichard/winpty#using-the-unix-adapter) suggests that any process spawned through `console.exe` will have a hidden console window, and it will convert between windows API calls to terminal esc codes.

If you don't want to build winpty, maybe python.exe(spawned with -i) is a good candidate.

equalsraf commented:
Tried to start :terminal, but the process seems to die immediately `Process exited -10`.

tarruda commented:
I have asked for some help in winpty issue tracker: https://github.com/rprichard/winpty/issues/63

equalsraf commented:
> I have asked for some help in winpty issue tracker

I'll have a look.

#810 now builds libvterm. This PR should be able to stand on its own now.

tarruda commented:
@equalsraf , can you test `python.exe -i`. If it works I say we should merge this and deal with winpty compatibility later.(As I [commented](https://github.com/rprichard/winpty/issues/63#issuecomment-173560365), it is possible to write another `console.exe` version that works when spawned from `CreateProcess` API)

tarruda commented on src/nvim/event/pty_process.h:
```diff
@@ -13,6 +18,27 @@ typedef struct pty_process {
+  Process process;
+  char *term_name;
+  uint16_t width, height;
+} PtyProcess;
```
I think it would be fine to just `#define PtyProcess LibuvProcess`/`#define pty_process_init libuv_process_init` and use constant values where `term_name`/`width`/`height` are queried. A clean way is to wrap all access to those fields in `pty_process` methods and define versions for `#ifndef FEAT_PTY_PROCESS` that return constants.

jalvesaq commented:
I can see the same problem described by @equalsraf (there is no output when I send lines of code to R):
<img width="379" alt="neovim_r_in_buffer" src="https://cloud.githubusercontent.com/assets/891655/12745845/315642b0-c97a-11e5-8be2-8789fd67f610.png">
It would be a great improvement over GVim if Neovim could run its terminal on Windows. I am waiting for this feature to stop the support for the Vim-R-plugin and keep maintaining only the Nvim-R plugin because I only use Neovim on Linux, but Windows users still can't run R in a Neovim terminal, which is the best way of running R controlled by Neovim.

rprichard committed:
Split pty_process into separate files for UNIX and Windows.

tarruda commented:
It seems we'll soon have `:term` working in windows, @rprichard made a [great contribution](https://github.com/rprichard/winpty/issues/63#issuecomment-188139209) and implemented `pty_process_win.c` by himself :smile:

jalvesaq commented:
Great news! I confirm that it works.

R does not start with `:terminal Rterm.exe`. When I do the Nvim-R command `<LocalLeader>rf` to start R, only `cmd.exe` is started. But after manually starting `Rterm` and doing `:call WaitNvimcomStart()` the communication with R works perfectly.

tarruda commented:
> R does not start with :terminal Rterm.exe. When I do the Nvim-R command <LocalLeader>rf to start R, only cmd.exe is started. But after manually starting Rterm and doing :call WaitNvimcomStart() the communication with R works perfectly.

cmd.exe is still hardcoded : https://github.com/equalsraf/neovim/blob/28f20057e1c6e88e8e7ca4806b8a8fccc354c6fa/src/nvim/event/pty_process_win.c#L58-L60

rprichard commented:
> cmd.exe is still hardcoded :

Yeah, I think it needs to convert the argv to a command-line using the [`CommandLineToArgvW`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776391(v=vs.85).aspx) convention.  There's an implementation in [libuv](https://github.com/libuv/libuv/blob/729d68596cc6289d60979f29f500cd46c99c6f74/src/win/process.c#L448) and in [winpty](https://github.com/rprichard/winpty/blob/9675a911a116387bc2bb135b8c2b4b1662993b6f/src/unix-adapter/main.cc#L188).

equalsraf commented:
Replaced the PR with @rprichard's first commit, as before it defines pty functions using the regular `libuv_process_` functions.

I made some minor changes (moved files into os/ and fixed a couple of lint errors).

jalvesaq commented:
Since I can't start R immediately with `termopen()` yet, I tried to start it with `jobsend("Rterm\n")`. However, it looks like the "\n" is not sent or cannot be used in cmd.exe. When I manually start `Rterm`, I can use `jobsend()` to send commands to the terminal and everything works as well as in Linux. For example, I can press `TAB` twice to get suggestions of completion in R. Perhaps this information is useful...

justinmk commented:
> However, it looks like the "\n" is not sent or cannot be used in cmd.exe.

@jalvesaq Try `\r\n`. Also try `jobsend()` with a list (e.g. `jobsend(1, ['Rterm'])`).

bfredl commented:
Needs to be `['Rterm', '']`

jalvesaq commented:
I tried `['Rterm', '']`. It doesn't work either.

I just tested with this C application:
```c
#include <stdio.h>
#include <string.h>

int main(int argc, char **argv)
{
    char b[128];

    printf("> ");
    while((fgets(b, 127, stdin))){
        if(strstr(b, "quit") == b)
            break;
        printf(" -->%s> ", b);
    }
    return 0;
}
```
And `jobsend` doesn't work with it too (it looks like the new line is not sent to the terminal). My conclusion is that Rterm has something special that makes `jobsend` functional.

jalvesaq commented:
@justinmk `jobsend(jobid, "Rterm\r\n")` works! Thank you!

equalsraf commented:
Ping, cleaned up and rebased.
